File: backend/app/schemas/resume.py
from pydantic import BaseModel, Field
from typing import Optional


class PersonalDetails(BaseModel):
    name: str = ""
    title: str = ""
    email: Optional[str] = None
    phone: Optional[str] = None
    location: Optional[str] = None
    linkedin: Optional[str] = None
    website: Optional[str] = None
    github: Optional[str] = None


class ExperienceItem(BaseModel):
    company: str
    title: str
    start_date: str
    end_date: str
    bullets: list[str] = Field(default_factory=list)


class ProjectItem(BaseModel):
    title: str
    description: str
    tech_stack: list[str] = Field(default_factory=list)
    status: Optional[str] = None
    url: Optional[str] = None


class EducationItem(BaseModel):
    institution: str
    degree: str
    year: str


class ResumeData(BaseModel):
    personal: PersonalDetails = Field(default_factory=PersonalDetails)
    summary: str = ""
    experience: list[ExperienceItem] = Field(default_factory=list)
    projects: list[ProjectItem] = Field(default_factory=list)
    education: list[EducationItem] = Field(default_factory=list)
    skills: list[str] = Field(default_factory=list)
    certifications: list[str] = Field(default_factory=list)
    awards: list[str] = Field(default_factory=list)


class ChangeExplanation(BaseModel):
    why: str
    confidence: float
    authenticity: str  # SAFE | STRETCH | UNVERIFIABLE
    interview_risk: str  # LOW | MEDIUM | HIGH


class SectionDiff(BaseModel):
    section: str
    original: str
    enhanced: str
    changes: list[ChangeExplanation] = Field(default_factory=list)
    accepted: Optional[bool] = None
    no_change_required: bool = False


class GapAnalysisOut(BaseModel):
    missing_skills: list[str] = Field(default_factory=list)
    missing_projects: list[str] = Field(default_factory=list)
    missing_certifications: list[str] = Field(default_factory=list)
    missing_frameworks: list[str] = Field(default_factory=list)


class ScoresOut(BaseModel):
    ai_readiness: float = 0
    ats_score: float = 0
    ai_terminology: float = 0
    project_evidence: float = 0
    technical_depth: float = 0
    measurable_achievements: float = 0
    band: str = "Beginner"


class SuggestedProject(BaseModel):
    title: str
    difficulty: str
    tech_stack: list[str]
    architecture: str
    roadmap: list[str]
    business_impact: str


class InterviewQuestionOut(BaseModel):
    id: str
    category: str
    question: str


class NextAction(BaseModel):
    title: str
    description: str
    type: str


class ResumeSuggestionOut(BaseModel):
    section: str
    title: str
    suggestion: str
    why: str


class PotentialCompanyOut(BaseModel):
    name: str
    role: str
    why_hiring: str
    fit_reason: str
    careers_hint: str = ""
    job_url: str = ""
    linkedin_url: str = ""
    match_score: int = 0
    search_keywords: str = ""


class ColdEmailOut(BaseModel):
    subject: str
    body: str


class OutreachOut(BaseModel):
    companies: list[PotentialCompanyOut] = Field(default_factory=list)
    cold_email: ColdEmailOut = Field(default_factory=lambda: ColdEmailOut(subject="", body=""))
    outreach_tip: str = ""


class DocumentUpdate(BaseModel):
    enhanced: str


class ResumeAnalysisOut(BaseModel):
    id: str
    status: str
    progress: int
    current_step: str
    original: ResumeData
    enhanced: ResumeData
    full_original: str = ""
    full_enhanced: str = ""
    diffs: list[SectionDiff]
    scores: ScoresOut
    gap_analysis: GapAnalysisOut
    suggested_projects: list[SuggestedProject]
    interview_questions: list[InterviewQuestionOut]
    next_actions: list[NextAction]
    suggestions: list[ResumeSuggestionOut] = Field(default_factory=list)
    potential_companies: list[PotentialCompanyOut] = Field(default_factory=list)
    cold_email: Optional[ColdEmailOut] = None
    outreach_tip: str = ""
    highlight_keywords: list[str] = Field(default_factory=list)
    ats_issues: list[str] = Field(default_factory=list)
    missing_keywords: list[str] = Field(default_factory=list)
    confidence_strengths: list[dict] = Field(default_factory=list)
    experience_gaps: list[dict] = Field(default_factory=list)
    original_filename: str = ""
    original_file_type: str = "pdf"


class SectionUpdate(BaseModel):
    enhanced: str
    accepted: Optional[bool] = None


class EnhancedDataUpdate(BaseModel):
    enhanced: ResumeData


class UploadResponse(BaseModel):
    id: str

File: frontend/src/types/resume.ts
export type AuthenticityLevel = "SAFE" | "STRETCH" | "UNVERIFIABLE";
export type InterviewRisk = "LOW" | "MEDIUM" | "HIGH";
export type ReadinessBand = "Beginner" | "Emerging" | "AI Aware" | "AI Ready";

export interface PersonalDetails {
  name: string;
  title: string;
  email?: string;
  phone?: string;
  location?: string;
  linkedin?: string;
  website?: string;
  github?: string;
}

export interface ExperienceItem {
  company: string;
  title: string;
  start_date: string;
  end_date: string;
  bullets: string[];
}

export interface ProjectItem {
  title: string;
  description: string;
  tech_stack?: string[];
  status?: string;
  url?: string;
}

export interface EducationItem {
  institution: string;
  degree: string;
  year: string;
}

export interface ResumeData {
  personal: PersonalDetails;
  summary: string;
  experience: ExperienceItem[];
  projects: ProjectItem[];
  education: EducationItem[];
  skills: string[];
  certifications?: string[];
  awards?: string[];
}

export interface ChangeExplanation {
  why: string;
  confidence: number;
  authenticity: AuthenticityLevel;
  interview_risk: InterviewRisk;
}

export interface SectionDiff {
  section: string;
  original: string;
  enhanced: string;
  changes: ChangeExplanation[];
  accepted?: boolean;
  no_change_required?: boolean;
}

export interface GapAnalysis {
  missing_skills: string[];
  missing_projects: string[];
  missing_certifications: string[];
  missing_frameworks?: string[];
}

export interface Scores {
  ai_readiness: number;
  ats_score: number;
  ai_terminology: number;
  project_evidence: number;
  technical_depth: number;
  measurable_achievements: number;
  band: ReadinessBand;
}

export interface SuggestedProject {
  title: string;
  difficulty: string;
  tech_stack: string[];
  architecture: string;
  roadmap: string[];
  business_impact: string;
}

export interface InterviewQuestion {
  id: string;
  category: "Behavioral" | "Technical" | "AI" | "Scenario-based";
  question: string;
}

export interface NextAction {
  title: string;
  description: string;
  type: "project" | "certification" | "learning";
}

export interface ResumeSuggestion {
  section: string;
  title: string;
  suggestion: string;
  why: string;
}

export interface PotentialCompany {
  name: string;
  role: string;
  why_hiring: string;
  fit_reason: string;
  careers_hint: string;
  job_url?: string;
  linkedin_url?: string;
  match_score?: number;
  search_keywords?: string;
}

export interface ColdEmail {
  subject: string;
  body: string;
}

export interface ResumeAnalysis {
  id: string;
  status: "pending" | "processing" | "completed" | "failed";
  progress: number;
  current_step: string;
  original: ResumeData;
  enhanced: ResumeData;
  full_original?: string;
  full_enhanced?: string;
  diffs: SectionDiff[];
  scores: Scores;
  gap_analysis: GapAnalysis;
  suggested_projects: SuggestedProject[];
  interview_questions: InterviewQuestion[];
  next_actions: NextAction[];
  suggestions?: ResumeSuggestion[];
  potential_companies?: PotentialCompany[];
  cold_email?: ColdEmail;
  outreach_tip?: string;
  highlight_keywords?: string[];
  ats_issues?: string[];
  missing_keywords?: string[];
  confidence_strengths?: { label: string; level: string }[];
  experience_gaps?: { label: string; level: string }[];
  original_filename?: string;
  original_file_type?: "pdf" | "docx";
}
